internal/build/resolver/user_builds.go

```diff
@@ -5,64 +5,47 @@ import (
 
 	"github.com/buildkite/cli/v3/internal/build"
-	"github.com/buildkite/cli/v3/internal/pipeline"
 	pipelineResolver "github.com/buildkite/cli/v3/internal/pipeline/resolver"
 	"github.com/buildkite/cli/v3/pkg/cmd/factory"
 	"github.com/buildkite/go-buildkite/v3/buildkite"
-	"golang.org/x/sync/errgroup"
 )
 
-// ResolveBuildFromCurrentBranch Finds the most recent build for the branch in the current working directory
-func ResolveBuildForCurrentUser(branch string, pipelineResolver pipelineResolver.PipelineResolverFn, f *factory.Factory) BuildResolverFn {
-	return func(ctx context.Context) (*build.Build, error) {
-		var pipeline *pipeline.Pipeline
-		var user *buildkite.User
+// ResolveBuildForUser Finds the most recent build for the user based on the provided options
+func ResolveBuildForUser(ctx context.Context, userInfo string, branch string, pipelineResolver pipelineResolver.PipelineResolverFn, f *factory.Factory) (*build.Build, error) {
 
-		// use an errgroup so a few API calls can be done in parallel
-		// and then we check for any errors that occurred
-		g, _ := errgroup.WithContext(ctx)
-		g.Go(func() error {
-			p, e := pipelineResolver(ctx)
-			if p != nil {
-				pipeline = p
-			}
-			return e
-		})
-		g.Go(func() error {
-			u, _, e := f.RestAPIClient.User.Get()
-			if u != nil {
-				user = u
-			}
-			return e
-		})
-		err := g.Wait()
-		if err != nil {
-			return nil, err
-		}
-		if pipeline == nil {
-			return nil, fmt.Errorf("failed to resolve a pipeline to query builds on.")
-		}
+	pipeline, err := pipelineResolver(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if pipeline == nil {
+		return nil, fmt.Errorf("failed to resolve a pipeline to query builds on")
+	}
+
+	opt := &buildkite.BuildsListOptions{
+		Creator: userInfo,
+		ListOptions: buildkite.ListOptions{
+			PerPage: 1,
+		},
+	}
 
-		builds, _, err := f.RestAPIClient.Builds.ListByPipeline(f.Config.OrganizationSlug(), pipeline.Name, &buildkite.BuildsListOptions{
-			Creator: *user.Email,
-			Branch:  []string{branch},
-			ListOptions: buildkite.ListOptions{
-				PerPage: 1,
-			},
-		})
-		if err != nil {
-			return nil, err
-		}
-		if len(builds) == 0 {
-			// we error here because this resolver is explicitly used so any case where it doesn't resolve a build is a
-			// problem
-			return nil, fmt.Errorf("failed to find a build for current user (email: %s)", *user.Email)
-		}
+	if len(branch) > 0 {
+		opt.Branch = []string{branch}
+	}
+
+	builds, _, err := f.RestAPIClient.Builds.ListByPipeline(f.Config.OrganizationSlug(), pipeline.Name, opt)
 
-		return &build.Build{
-			Organization: f.Config.OrganizationSlug(),
-			Pipeline:     pipeline.Name,
-			BuildNumber:  *builds[0].Number,
-		}, nil
+	if err != nil {
+		return nil, err
 	}
+	if len(builds) == 0 {
+		// we error here because this resolver is explicitly used so any case where it doesn't resolve a build is a
+		// problem
+		return nil, fmt.Errorf("failed to find a build for current user (%s)", userInfo)
+	}
```

Comment:
I wonder if this applies now with these other resolvers? I suppose if the user doesn't have any builds its not really an error, we could just show a helpful message without it being an error

Reply:
I guess thats right, and for consistency too as we do the same for pipelines resolver. I'll update this to not error when no builds are found

Reply:
actually its probably more nuanced than that. the `--user` resolver might want to error if that user doesnt exist vs. if the user does exist but doesnt have any builds

Reply:
@jradtilbrook , i will raise a different PR to implement this as it will have to be the same behavior for all build resolvers and ensure that this is handled as expected.

Reply:
if the `user` does not exist, or is invalid the api returns an error, which is also returned by the resolver. So that should be covered.